kratos/geometries/geometry.h

```diff
@@ -901,29 +906,38 @@ class Geometry : public PointerVector<TPointType>
 
         unsigned int maxiter = 1000;
 
-        for ( unsigned int k = 0; k < maxiter; k++ )
-        {
-            CurrentGlobalCoords = ZeroVector( 3 );
-            GlobalCoordinates( CurrentGlobalCoords, rResult );
-            noalias( CurrentGlobalCoords ) = rPoint - CurrentGlobalCoords;
-            InverseOfJacobian( J, rResult );
-            noalias( DeltaXi ) = prod( J, CurrentGlobalCoords );
-            noalias( rResult ) += DeltaXi;
-
-//            if ( MathUtils<double>::Norm3( DeltaXi ) > 30 )
-            if ( norm_2( DeltaXi ) > 30 )
-            {
-                break;
+        if(LocalSpaceDimension() == 2) {
+            for ( unsigned int k = 0; k < maxiter; k++ ) {
+                CurrentGlobalCoords = ZeroVector( 3 );
+                GlobalCoordinates( CurrentGlobalCoords, rResult );
+                noalias( CurrentGlobalCoords ) = rPoint - CurrentGlobalCoords;
+                InverseOfJacobian( J, rResult );
+                DeltaXi[0] = J(0,0) * CurrentGlobalCoords(0) + J(0,1) * CurrentGlobalCoords(1);
```

Comment:
now this only works in 2D but not in 3D... 

i would do this as

    DeltaXi.clear();
     for(unsigned int i=0; i<WorkingSpaceDimension(); ++i
           for(unsigned int j=0; k<WorkingSpaceDimension(); ++j)
                DeltaXi[i] += J(i,j)*CurrentGlobalCoords[j];

Reply:
It does work in 3d because doesn't use that code, I just separated the cases. I assume the prod from boost had some kind of in-built optimization for matrices and I didn't wanted to touch it. If is not the case I can implement what you said.